Implement Promise.withResolvers ES2024 feature (fixes #1544)

anivar opened the pull request:
## Summary

Implements the ES2024 Promise.withResolvers() static method as specified in the ECMAScript 2024 specification.

## Background

Promise.withResolvers() provides a cleaner alternative to the traditional Promise constructor pattern when you need access to resolve/reject functions outside the executor:

**Before (traditional pattern):**
```javascript
let resolve, reject;
const promise = new Promise((res, rej) => {
  resolve = res;
  reject = rej;
});
```

**After (with Promise.withResolvers):**
```javascript
const { promise, resolve, reject } = Promise.withResolvers();
```

## Implementation Details

- Added withResolvers as a static method on the Promise constructor
- Uses the existing Capability class to ensure proper promise construction
- Returns a plain object with promise, resolve, and reject properties
- Properties are writable, enumerable, and configurable (as per spec)
- Supports subclassing correctly (uses thisObj for construction)

## Tests

Added comprehensive unit tests to PromiseTest.java covering:
- Basic functionality and return types
- Promise instanceof validation  
- Object structure and property attributes
- Resolve and reject functionality

All existing Promise tests continue to pass.

Fixes #1544

rbri commented on rhino/src/main/java/org/mozilla/javascript/NativePromise.java:
```diff
@@ -313,6 +315,24 @@ private static Object any(Context cx, Scriptable scope, Scriptable thisObj, Obje
 
+    // Promise.withResolvers
+    private static Object withResolvers(Context cx, Scriptable scope, Scriptable thisObj, Object[] args) {
+        if (!ScriptRuntime.isObject(thisObj)) {
```
maybe you can add a test for this also

anivar commented:
Thanks for the review @rbri\! 

I'll separate out the ISO date parsing changes - looks like those got mixed in accidentally. Let me clean up this PR to focus solely on the Promise.withResolvers implementation.

Regarding tests - I've already included comprehensive tests in PromiseTest.java covering:
- Basic functionality and return types
- Promise instanceof validation  
- Object structure and property attributes
- Resolve and reject functionality

I'll push an update shortly with just the Promise.withResolvers changes.

anivar commented:
@rbri I've addressed your feedback:

1. ✅ Removed the ISO date parsing changes - PR now only contains Promise.withResolvers implementation
2. ✅ Added test for the non-object TypeError case you pointed out

The PR now includes:
- Implementation of Promise.withResolvers in NativePromise.java
- Comprehensive tests in PromiseTest.java including the TypeError test

Ready for review\!

rbri commented:
@anivar hopefully you are not frustrated from the long way to get this merged. We really appreciate your effort.

The checks are failing because of spotless. Simply run

    ./gradlew :rhino:spotlessApply

to fix that.

anivar commented:
@rbri Thanks for the encouragement\! I'm having issues with my local Java setup to run spotless. Could you help run `./gradlew :rhino:spotlessApply` on the PR? I don't want to delay the merge over a formatting issue.

anivar commented:
Update: My Java installation seems to be having issues. The PR is functionally complete and all tests are passing. Would you mind running the spotless formatter on your end? Thanks!

anivar commented:
Fixed! Java setup is working now and spotless formatting has been applied.

rbri commented:
@anivar  maybe one last thing :-D

Please update the test262.properties file (see https://github.com/mozilla/rhino/tree/master/tests#updating-the-test262properties-file). I guess your improvement makes some more tests green here.
